openmeter/ent/schema/subscription.go

```diff
@@ -0,0 +1,141 @@
+package schema
+
+import (
+	"entgo.io/ent"
+	"entgo.io/ent/dialect"
+	"entgo.io/ent/schema/edge"
+	"entgo.io/ent/schema/field"
+	"entgo.io/ent/schema/index"
+
+	"github.com/openmeterio/openmeter/openmeter/subscription"
+	"github.com/openmeterio/openmeter/pkg/currencyx"
+	"github.com/openmeterio/openmeter/pkg/datex"
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+)
+
+type Subscription struct {
+	ent.Schema
+}
+
+func (Subscription) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+		entutils.CadencedMixin{},
+	}
+}
+
+func (Subscription) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("plan_key").NotEmpty().Immutable(),
+		field.Int("plan_version").Min(1).Immutable(),
+		field.String("customer_id").NotEmpty().Immutable(),
+		field.String("currency").GoType(currencyx.Code("")).MinLen(3).MaxLen(3).NotEmpty().Immutable(),
+	}
+}
+
+func (Subscription) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "customer_id"),
+	}
+}
+
+func (Subscription) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("entitlements", SubscriptionEntitlement.Type),
+		edge.From("customer", Customer.Type).Field("customer_id").Ref("subscription").Immutable().Unique().Required(),
+		edge.To("phases", SubscriptionPhase.Type),
+	}
+}
+
+type SubscriptionPhase struct {
+	ent.Schema
+}
+
+func (SubscriptionPhase) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (SubscriptionPhase) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("subscription_id").NotEmpty().Immutable(),
+		field.String("key").NotEmpty().Immutable(),
+		field.String("name").NotEmpty(),
+		field.String("description").Optional().Nillable(),
+		field.Time("active_from").Immutable(),
```

Comment:
don't we need an optional active_to? Or that coming from rate cards somehow? If yes why do we need active_from?

Reply:
By the current spec, a subscription cannot have a state where there isn't an active phase, unless the subscription itself is inactive (cancelled etc...) it's just a simple way of guaranteeing it